### ISSUE TITLE
Maize genome in riboviz format

### ISSUE THREAD

**Issue by NastiaSkuba** [riboviz/RiboViz#485]
Hi all,

I would like to use your package for maize ribo-seq dataset and currently I am trying to convert the maize genome data into riboviz-friendly format.

I have got the B73 GRAMENE v4 genome with v41 annotation from here:
https://nov2020-plants.ensembl.org/Zea_mays/Info/Index

I have genome, cDNA and CDS fasta, but only genome gff.

I have tried `create_filez_for_riboviz.R` script, but it gives me the next error:
```
[1] "Reading in Genome..."
[1] "Done"
[1] "Reading in GFF3..."
[1] "Done"
[1] "Adding Buffer to genome..."
[1] "Done"
[1] "Filtering GFF3 by provided criteria..."
[1] "Done"
[1] "Removing trailing values on gene ids..."
[1] "Done"
[1] "Creating riboviz-style CDS..."
Error: logical subscript contains NAs
Execution halted
```

I have installed as much as possible packages from conda. 
```
> sessionInfo()
R version 3.6.1 (2019-07-05)
Platform: x86_64-conda_cos6-linux-gnu (64-bit)
Running under: Ubuntu 22.04.4 LTS

Matrix products: default
BLAS/LAPACK: /home/qbase/miniconda3/envs/riboviz-env/lib/R/lib/libRblas.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=de_DE.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=de_DE.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=de_DE.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=de_DE.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    parallel  stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] rhdf5_2.30.0         stringr_1.4.0        rtracklayer_1.46.0  
 [4] GenomicRanges_1.38.0 GenomeInfoDb_1.22.1  Biostrings_2.54.0   
 [7] XVector_0.26.0       IRanges_2.20.2       S4Vectors_0.24.4    
[10] BiocGenerics_0.32.0  argparse_2.0.3      

loaded via a namespace (and not attached):
 [1] magrittr_1.5                zlibbioc_1.32.0            
 [3] GenomicAlignments_1.22.1    BiocParallel_1.20.1        
 [5] lattice_0.20-38             R6_2.4.0                   
 [7] tools_3.6.1                 SummarizedExperiment_1.16.1
 [9] grid_3.6.1                  Biobase_2.46.0             
[11] matrixStats_1.3.0           Matrix_1.2-17              
[13] GenomeInfoDbData_1.2.2      Rhdf5lib_1.8.0             
[15] bitops_1.0-7                RCurl_1.98-1.14            
[17] stringi_1.4.3               DelayedArray_0.12.3        
[19] compiler_3.6.1              Rsamtools_2.2.3            
[21] XML_3.99-0.3                jsonlite_1.6  
```

Do you have any suggestions how to get the script working?
Or is there any ways to get cDNA/CDS gff, if I already have fasta?

I would be grateful for any help!

**Comment by acope3** [comment]
Hi @NastiaSkuba,

Apologies for taking a while to get back to you. I am currently at a conference. Have a look at the branch `create_riboviz_style_gff_acope3`. There is the R script `rscripts/create_files_for_riboviz.R`. This file is intended to take a GFF3/GTF and a FASTA file (with chromosome sequences) like you might download from a database and convert it to a riboviz style CDS FASTA and GFF3 file. It is still being tested, so I definitely would recommend doing some sanity checks. Please let me know if you have any questions!

**Comment by NastiaSkuba** [comment]
I have tried the R script `rscripts/create_files_for_riboviz.R` with `--seq_id` parameter set to `ID`
The code is executed for around 10 min, but then I am getting the next error:

```
[1] "Reading in Genome..."
[1] "Done"
[1] "Reading in GFF3..."
[1] "Done"
[1] "Adding Buffer to genome..."
[1] "Done"
[1] "Filtering GFF3 by provided criteria..."
[1] "Done"
[1] "Removing trailing values on gene ids..."
[1] "Done"
[1] "Creating riboviz-style CDS..."
Error in .subset_by_GenomicRanges(x, i) : 
  when subsetting by a GenomicRanges subscript, the names of the object
  to subset must contain the seqnames of the subscript
Calls: convertGenomeToCDSFile -> [ -> [ -> .subset_by_GenomicRanges
Execution halted
```
A new fasta file was NOT created

Here is the example of .gff file I have:

```
1	gramene	gene	44289	49837	.	+	.	ID=gene:Zm00001d027230;biotype=protein_coding;description=Mitochondrial transcription termination factor family protein;gene_id=Zm00001d027230;logic_name=maker_gene
1	gramene	mRNA	44289	49837	.	+	.	ID=transcript:Zm00001d027230_T001;Parent=gene:Zm00001d027230;biotype=protein_coding;transcript_id=Zm00001d027230_T001
1	gramene	five_prime_UTR	44289	44350	.	+	.	Parent=transcript:Zm00001d027230_T001
1	gramene	exon	44289	44947	.	+	.	Parent=transcript:Zm00001d027230_T001;Name=Zm00001d027230_T001.exon1;constitutive=1;ensembl_end_phase=0;ensembl_phase=-1;exon_id=Zm00001d027230_T001.exon1;rank=1
1	gramene	CDS	44351	44947	.	+	0	ID=CDS:Zm00001d027230_P001;Parent=transcript:Zm00001d027230_T001;protein_id=Zm00001d027230_P001
```

What can be reason of such an error?

**Comment by acope3** [comment]
Hi @NastiaSkuba, can you please share with me the exact command you used for calling the script?

**Comment by NastiaSkuba** [comment]
```
Rscript --vanilla rscripts/create_files_for_riboviz.R \
-i ../ribo-organisms/mays/Zea_mays.B73_RefGen_v4.dna.toplevel.fa \
-g ../ribo-organisms/mays/Zea_mays.B73_RefGen_v4.41. \
--out_dir ../ribo-organisms/mays/riboviz-style \
--seq_id ID \
--exons_preordered
```

**Comment by acope3** [comment]
Hi @NastiaSkuba I apologize for the long delays between communications. Funding for this project has long expired, so I'm pretty much the only one trying to maintain things. I am working on this a bit today, so hopefully will have something for you soon. A few things:

(1) If you need this ASAP, you can check out the `rmarkdown/script_for_transcript_annotation.Rmd`, specifically the code for the human dataset. This should allow you to use the cDNA files from Ensembl, but you will need to define principal isoforms for each gene. Unfortunately, I don't think the Appris database has these for Z. mays.  

(2) Another advantage of the R markdown file is it allows you to use the annotated UTRs. The R script you have been using appends a buffer region (default to 250 nucleotides both up and downstream of the CDS) -- this is useful if the UTR annotations are not generally reliable. The R script is an attempt to generalize the R markdown file, but it needs to be modified to use annotated UTRs, if the user so chooses. If you don't trust the UTR annotations for Z. mays, then probably better to use a buffer, in which case we will want to continue with the R-script. You will still need a list of principal isoforms.

(3) Regardless, it would be good to add the riboviz-style FASTA and GFF3 files to our [example-datasets repository](https://github.com/riboviz/example-datasets) if it is okay with you.

**Comment by acope3** [comment]
Hi @NastiaSkuba,

I ran this locally on a subset of transcripts and chromosomes. It outputs a FASTA and GFF3 file. I noticed your code output had the line "Removing trailing values on gene ids...". My local version had a different default for the relevant command line argument compared to the version on branch [create_riboviz_style_cds_gff_acope3-278](https://github.com/riboviz/riboviz/tree/create_riboviz_style_cds_gff_acope3-278). I have updated this code to have a `NULL` as the default. Can you please try again on the full set of genes to see if this resolves your error? 

Also, double-check if the exons are preordered. Based on the GFF3 file I downloaded, it looks like the exons are not preordered.

**Comment by NastiaSkuba** [comment]
Thank you for your help!

After update everything worked. Actually, the "Removing trailing values on gene ids..." was not the only one problem. When I changed only that line, I got another error message. But when I updated the whole document, the code worked.
Also transcripts are marked with protein names. It seems strange to me, but does not affect the execution so far.

For maize only about half of the genes have more or less annotated UTRs, and quality of this annotation is not clear. So it is debatable, if one should trust UTR annotation. Nevertheless, I surely will upload the files I will create to the example-datasets repository.
